### DISCUSSION TITLE
Problem with MkDoxy configuration verification - new schema

### DISCUSSION THREAD

**Discussion by JakubAndrysek** [mkdocs/mkdocs#3744] [General]
Hi, I am [MkDoxy](https://mkdoxy.kubaandrysek.cz/) plugin developer. I have found time to work on the plugin and update the structure and main core.
Based on the configured class, I have found the "new" option for [specifying config_scheme](https://www.mkdocs.org/dev-guide/plugins/#subclassing-config-to-specify-the-config-schema). This is a great feature, and I would like to use it in my plugin.
But three years ago, when I was developing the plugin, I created config_scheme with keys that [contain hyphens](https://github.com/JakubAndrysek/MkDoxy/blob/1702d0f13cb7cac74ba295e12f299934aa9c292d/mkdocs.yml#L63-L89) ("full-doc", "src-dirs", ...).

Now, I would like to use the new feature. I think the best way is to [migrate](https://github.com/JakubAndrysek/MkDoxy/blob/core-update/mkdoxy/migration.py#L44) the existing config_scheme to the new format.
This tool is starting to be more and more popular, and I would like to keep them working also with the old schema (until developers update their configuration).

My plan was to use both schemas, but it seems impossible. Keys in 

 I thought that when I call `on_startup` method, I can check if the new schema is used and if not, I can convert it to the new format. 
But it is impossible because the `mkdocs.yaml` is checked before calling this method.

Does anyone have an idea how to solve this problem?
Is the easiest way to stay with the old schema and wait for developers to update their configuration?

Have a nice day!
Jacob

**Comment by pawamoy** [comment]
Hi Jacob!

IIUC you want to use the new way of declaring your plugin configuration, but this new way doesn't allow to use hyphens in option names since options are declared as class variables, and Python variables cannot use hyphens in their name. Therefore, you cannot switch cleanly to the new way of declaring options because you won't be able to support the old option names during a transition period.

If I got this right, what I suggest then is that  you keep using the old way for declaring your options, and that you have a deprecation period where both names (with hyphens and with underscore) are supported, while the old names are deprecated and emit deprecation warnings. After some time, your users should have upgraded to the new names, and you can "safely" switch to the new way for declaring options.

**Reply by JakubAndrysek** [reply]
Thank you for your reply. I will update it as you described by using the deprecation warning. 

I will be glad to find the recommendation to use an underscore as a separator in a longer config key (better for future migration to the new schema if someone starts with the old one).
I know that there is the `option_name` [example](https://www.mkdocs.org/dev-guide/plugins/#using-plugins), but I would like to see the info in the text.

Thanks

**Reply by JakubAndrysek** [reply]
Same as that the new validation is processed before the `on_startup` methot is called.
Right now the documentation contains:
  > The `startup` event runs once at the very beginning of an `mkdocs` invocation.

Which sounds like "right after" executing `mkdocs`

**Reply by pawamoy** [reply]
Yes, the configuration file is always loaded before anything else, because everything else needs the configuration file to be loaded. That shouldn't prevent you from using `on_config` to check the options provided in the config file though, and emit deprecation warnings where necessary.

**Reply by kamilkrzyskow** [reply]
You can always check the source to see what is happening, the config has to loaded before everything else to attach all of the hook functions etc. from plugins to later call `on_startup` on the `cfg` object.
https://github.com/mkdocs/mkdocs/blob/2ed8798835521ae8f9da07280dd93d92944f8372/mkdocs/__main__.py#L271-L286

I also think that the deprecation approach would be for the best. However, I think it's possible to turn off the validation for a PluginConfig during load and defer it to the `on_config`, and I think it's possible to support both `full-doc` and `full_doc` without deprecation, of course not as `config.full-doc` as this would break Python, but instead make `config.full_doc` load `config["full-doc"]`, and you'd have to sync it yourself. Probably possible, but also probably a nightmare to support and maintain 😅
